### PR TITLE
add privileges for kIbana_system user to serve cloud security posture…

### PR DESCRIPTION
**Pull request by CohenIdo**
Add privileges for `kibana_system` user to allow support in Cloud Security Posture's feature.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (joshdover) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:788]: Any reason why the names of these indices can't match the [data stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme) we use for all other integration indices?

We can still make them hidden, if desired, see https://github.com/elastic/package-spec/issues/266

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (joshdover) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:790]: Do we need `delete`? I don't see it as a requirement for the transform APIs nor do I expect Kibana to be deleting documents from these indices

Comment (joshdover) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: @elastic/ml-data (I hope this is the right team) are our docs out of date here? I'm surprised that installing a transform is attempting a delete by query on the destination index and that the docs don't indicate this. I'm also surprised this issue doesn't affect our other transforms installed by Fleet?

> ```
> task encountered irrecoverable failure: org.elasticsearch.ElasticsearchSecurityException: action [indices:data/write/delete/byquery] is unauthorized for user [kibana_system] with roles [kibana_system] on indices [cloud_security_posture-findings_latest], this action is granted by the index privileges [delete,write,all]
> ```

That error seems to indicate that they need one of these `[delete,write,all]` privileges, but not all of them. Can we test what the minimum required here is?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (joshdover) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:808]: Do we still need these?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kfirpeled) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Thanks josh for calling this out
Since these are indices and not datastreams we assumed it would be more fit not to stick with the same naming scheme. And since `latest` transform actually deletes documents from the index and is not append-only kind of index like datastream it can't be a datastream.

I saw this https://github.com/elastic/kibana/issues/121118 as well and wondering what you think is best fit here
Because that raises some questions:
1. what is the type we should use when Transform latest is used?
2. what is the type we should use when Transform pivot is used?
3. how can we support namespaces? and if we should address that with an adhoc solution

Comment (joshdover) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: > Since these are indices and not datastreams we assumed it would be more fit not to stick with the same naming scheme. And since latest transform actually deletes documents from the index and is not append-only kind of index like datastream it can't be a datastream.

Though they aren't data streams, users are going to be granted index access based on index patterns and I think we'd want a pattern to match both the source data streams and the transform destination indices so that users can easily configure access to this data.

> 1\. what is the type we should use when Transform latest is used?

Good question, IMO `logs` makes sense for this use case, and it also matches the source data stream. This could be a good discussion to open up in the issue related to adding official support for `transforms` in the package-spec: elastic/package-spec#23

> 2\. what is the type we should use when Transform pivot is used?

Like above, I think we can match the source data stream's type

> 3\. how can we support namespaces? and if we should address that with an adhoc solution

You'd have to install the transform yourself, but IMO it's probably not worth supporting namespaces in your MVP. I think we should tackle namespace-specific transforms at the Fleet package level and this should be discussed as part of elastic/package-spec#23 as well.

Comment (kfirpeled) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: thanks a lot josh :) so we will name it
`logs-cloud_security_posture.findings_latest-default` and `logs-cloud_security_posture.benchmark_scores-default`
So that it would fit our planned renaming of our datastreams and future namespace support (currently we would only support default anyway)

**Comment by ywangd** [comment]
Ping @elastic/kibana-security for review

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (kfirpeled) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Hey @joshdover , so we encounter an issue when attempting to create and index that starts with `logs`.

`[2022-03-24T11:56:18.423+02:00][ERROR][plugins.cloudSecurityPosture] illegal_argument_exception: cannot create index with name [logs-cloud_security_posture.findings_latest-default], because it matches with template [logs] that creates data streams only, use create data stream api instead`

since it can't be a datastream, how can we proceed from here? obviously it can't be a datastream because the Transform writes and deletes documents from that index.

Comment (joshdover) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Hmm shouldn't there be an index template for this index as part of your package? It's important that the index template's priority is high enough to override the default `logs` template

Comment (CohenIdo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: @joshdover, thanks for that input, it worked.

@kfirpeled, I change the indices names to:
`logs-cloud_security_posture.findings_latest-default` and `logs-cloud_security_posture.scores-default`

**Review by kfirpeled — COMMENTED** [review]
namings lgtm

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (Winterflower) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Hi @joshdover - I think @elastic/ml-ui would be more help here

**Review by jportner — APPROVED** [review]
Kibana Security Review --

Thanks for tagging us, LGTM for posterity

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: @joshdover we need delete by query. This is for the new data retention policies that transforms can create. Our docs our out of date if delete is not included.

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: I will create an ES issue to capture this inaccuracy and adding a check early on to verify security (like we do for write/read).

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: https://github.com/elastic/elasticsearch/issues/85409 for tracking our missing validation and documentation.

Comment (CohenIdo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Thank you all, I added the `delete` privilege.

**Review by joshdover — APPROVED** [review]
LGTM from a Fleet perspective, still need approval from @elastic/es-security

**Comment by sophiec20** [comment]
For full disclosure - We are just starting to work on 2 more Security packages (host risk score and beaconing) which contain multiple transforms, and we anticipate more in the near future. 

Are we sure that extending the privileges of `kibana_system` is how we want to achieve this? In this case, we are using `kibana_system` as a type of service account. This is just about acceptable when transforming internal data (as in the security posture use case) but if the use case requires source data to be a data stream and as the number of transforms grows, then we should consider a longer term solution. Due to the split nature of the elasticsearch and kibana security model there are no easy options here, but I wanted to raise awareness that this is not a one-off request on `kibana_system`.

**Review by albertzaharovits — APPROVED** [review]
LGTM
My personal preference, not a requirement, is that system generated data go into index names that start with `.`.

**Comment by joshdover** [comment]
> Are we sure that extending the privileges of `kibana_system` is how we want to achieve this? In this case, we are using `kibana_system` as a type of service account. This is just about acceptable when transforming internal data (as in the security posture use case) but if the use case requires source data to be a data stream and as the number of transforms grows, then we should consider a longer term solution. Due to the split nature of the elasticsearch and kibana security model there are no easy options here, but I wanted to raise awareness that this is not a one-off request on `kibana_system`.

@sophiec20 Agreed, this is a long-term solution we can keep sustaining. Please see https://github.com/elastic/package-spec/issues/293 for discussion of other options. AFAIK no option has been proposed that wouldn't involve having Elasticsearch implement a package install API. If transforms shipped in packages is going to be a growing feature we need to start having more discussions with Elasticsearch about implementing this feature.

cc @ruflin

**Pull request merged**
